File: clients/kotlin/src/main/kotlin/streamforge/GrpcTransport.kt
package streamforge

import com.google.protobuf.Empty
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.Metadata
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map
import streamforge.v1.GetTableRowsRequest
import streamforge.v1.IngestRequest
import streamforge.v1.IngestServiceGrpcKt
import streamforge.v1.StreamServiceGrpcKt
import streamforge.v1.SubscribeTableRequest
import streamforge.v1.TableServiceGrpcKt

private val AUTH_KEY: Metadata.Key<String> = Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER)

/** Result of one `IngestService.Ingest` bidi push (design doc §3.1: real HTTP/2 backpressure --
 * the server does not ack until the push is admitted -- instead of REST's `retry_after_ms`
 * guess). Mirrors `StreamForge.Abstractions.IngestOutcome` (`IngestAck` in streamforge.proto). */
data class IngestAckResult(
    val outcome: String,
    val accepted: Int,
    val dropped: Int,
    val invalid: Int,
    val retryAfterMs: Int,
    val error: String,
    val rowErrors: List<String>,
    val duplicate: Int,
    val replayed: Boolean,
)

/**
 * Tier 1 gRPC transport (design doc §3): `StreamService.SubscribeTable` for deltas,
 * `TableService.Rows/List` for the catalog and snapshot, bidi `IngestService.Ingest` for pushes.
 * One insecure h2c channel, prior knowledge -- matching how the engine actually runs from source
 * (§3.2's `--urls` trap: it must be started with `--Http:Port`/`--Grpc:Port`, never `--urls`, or
 * no gRPC port is bound at all).
 *
 * Every gRPC service carries `[Authorize(Policy = "Viewer")]`, so an `authorization: Bearer <jwt>`
 * metadata entry is all that's needed -- [tokenProvider] is `AuthClient::token`, shared with the
 * REST side so both transports mint/refresh the same JWT (design doc §3.1).
 */
class GrpcTransport(target: String, private val tokenProvider: suspend () -> String) : TableTransport {
    override val name = "grpc"

    private val channel: ManagedChannel = ManagedChannelBuilder.forTarget(target).usePlaintext().build()
    private val tableStub = TableServiceGrpcKt.TableServiceCoroutineStub(channel)
    private val streamStub = StreamServiceGrpcKt.StreamServiceCoroutineStub(channel)
    private val ingestStub = IngestServiceGrpcKt.IngestServiceCoroutineStub(channel)

    private suspend fun authHeaders(): Metadata = Metadata().apply { put(AUTH_KEY, "Bearer ${tokenProvider()}") }

    override fun close() {
        channel.shutdownNow()
    }

    /** Proves the channel AND the JWT actually work -- used by `StreamForge.connect(transport =
     * AUTO)` to decide whether gRPC is reachable before committing to it (design doc: `AUTO`
     * "tries gRPC, falls back to SignalR, and logs which one it got"). */
    suspend fun probe() {
        tableStub.list(Empty.getDefaultInstance(), authHeaders())
    }

    private suspend fun resolveTableId(tableName: String): String {
        val resp = tableStub.list(Empty.getDefaultInstance(), authHeaders())
        return resp.tablesList.firstOrNull { it.name == tableName }?.id
            ?: throw StreamForgeError("no such table '$tableName'")
    }

    override suspend fun snapshot(tableName: String, limit: Int): Pair<List<RowDelta>, Long> {
        val id = resolveTableId(tableName)
        val request = GetTableRowsRequest.newBuilder().setId(id).setLimit(limit).build()
        val response = tableStub.rows(request, authHeaders())
        return response.rowsList.map { RowDelta(it.row.toRow(), it.weight) } to response.seq
    }

    override suspend fun subscribeTable(tableName: String): Flow<DeltaBatch> {
        // Metadata (the bearer token) is fetched HERE, as part of this suspend call, not lazily
        // inside a `flow {}` builder -- see [TableTransport.subscribeTable]'s doc: a cold Flow
        // that defers even cheap prep work to collection time can let the caller's REST snapshot
        // race ahead of the subscription actually being registered. gRPC's own stream setup
        // (`streamStub.subscribeTable`) stays a plain non-suspend call returning a Flow, same as
        // grpc-kotlin generates it -- fetching the token first is what matters here, since the
        // token itself refreshes on every (re)subscribe, mirroring live.py's "every reconnect
        // starts fresh".
        val headers = authHeaders()
        val request = SubscribeTableRequest.newBuilder().setName(tableName).build()
        return streamStub.subscribeTable(request, headers).map { batch ->
            DeltaBatch(batch.deltasList.map { RowDelta(it.row.toRow(), it.weight) }, batch.seq)
        }
    }

    /** One request, one ack, over a fresh bidi stream -- gets real backpressure semantics from
     * the bidi RPC without holding a stream open across calls (same simplification the Python
     * client makes in `_grpc.py`'s `ingest()`: a long-lived streaming session across many pushes
     * is future work, not needed for `push()`'s surface). */
    suspend fun ingest(sourceName: String, rows: List<Row>, idempotencyKey: String?, partial: Boolean): IngestAckResult {
        val request = IngestRequest.newBuilder()
            .setSourceName(sourceName)
            .addAllRows(rows.map { it.toStruct() })
            .setPartial(partial)
            .setIdempotencyKey(idempotencyKey ?: "")
            .build()
        val ack = ingestStub.ingest(flowOf(request), authHeaders()).first()
        return IngestAckResult(
            outcome = ack.outcome.name,
            accepted = ack.accepted,
            dropped = ack.dropped,
            invalid = ack.invalid,
            retryAfterMs = ack.retryAfterMs,
            error = ack.error,
            rowErrors = ack.rowErrorsList,
            duplicate = ack.duplicate,
            replayed = ack.replayed,
        )
    }
}

File: clients/dotnet/StreamsForge.Client/GrpcTransport.cs
using System.Runtime.CompilerServices;
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;
using Grpc.Net.Client;
using StreamsForge.Host.Grpc.V1;

namespace StreamsForge.Client;

/// <summary>
/// Tier 1 gRPC transport: <c>StreamService.SubscribeTable</c> for deltas, <c>TableService.Rows</c>
/// for the snapshot, <c>TableService.List</c> for id resolution and as the Auto-connect probe, and
/// the bidi <c>IngestService.Ingest</c> for pushes. One insecure h2c channel, prior knowledge -- no
/// TLS negotiation, matching how the engine is actually run from source (the <c>--urls</c> trap:
/// starting the host with <c>--urls</c> trips its guard so no gRPC port is bound at all).
///
/// Row payloads travel as <c>google.protobuf.Struct</c>; <see cref="RowCodec"/> is the whole
/// "typing" story -- see its class doc for the Struct-number precision hazard.
/// </summary>
internal sealed class GrpcTransport : ITransport, IAsyncDisposable
{
    public string Name => "grpc";

    private readonly GrpcChannel _channel;
    private readonly TableService.TableServiceClient _tables;
    private readonly StreamService.StreamServiceClient _stream;
    private readonly IngestService.IngestServiceClient _ingest;
    private readonly Func<CancellationToken, ValueTask<string>> _tokenProvider;

    public GrpcTransport(string target, Func<CancellationToken, ValueTask<string>> tokenProvider)
    {
        // h2c: the engine serves gRPC in cleartext locally (--Grpc:Port), so the client must ask
        // for HTTP/2 over plain HTTP explicitly -- .NET's SocketsHttpHandler otherwise refuses to
        // negotiate HTTP/2 without TLS ("prior knowledge" h2c).
        AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);
        _channel = GrpcChannel.ForAddress($"http://{target}", new GrpcChannelOptions
        {
            HttpHandler = new SocketsHttpHandler(),
        });
        _tables = new TableService.TableServiceClient(_channel);
        _stream = new StreamService.StreamServiceClient(_channel);
        _ingest = new IngestService.IngestServiceClient(_channel);
        _tokenProvider = tokenProvider;
    }

    private async Task<Metadata> AuthMetadataAsync(CancellationToken ct)
    {
        var token = await _tokenProvider(ct).ConfigureAwait(false);
        return new Metadata { { "authorization", $"Bearer {token}" } };
    }

    // ---- catalog (id resolution + the Auto-connect "does the channel and JWT actually work" probe) ----

    public async Task<IReadOnlyList<TableSummary>> ListTablesAsync(CancellationToken ct)
    {
        var md = await AuthMetadataAsync(ct).ConfigureAwait(false);
        var resp = await _tables.ListAsync(new Empty(), md, cancellationToken: ct).ConfigureAwait(false);
        return resp.Tables.Select(t => new TableSummary(t.Id, t.Name)).ToList();
    }

    private async Task<string> ResolveTableIdAsync(string name, CancellationToken ct)
    {
        foreach (var t in await ListTablesAsync(ct).ConfigureAwait(false))
        {
            if (t.Name == name) return t.Id;
        }
        throw new StreamsForgeException($"no such table '{name}'");
    }

    // ---- ITransport ----

    public async Task<(IReadOnlyList<RowDelta> Rows, long Seq)> SnapshotAsync(string table, int limit, CancellationToken ct)
    {
        var id = await ResolveTableIdAsync(table, ct).ConfigureAwait(false);
        var md = await AuthMetadataAsync(ct).ConfigureAwait(false);
        var resp = await _tables.RowsAsync(new GetTableRowsRequest { Id = id, Limit = limit }, md, cancellationToken: ct)
            .ConfigureAwait(false);
        var rows = resp.Rows.Select(r => new RowDelta(RowCodec.FromStruct(r.Row), r.Weight)).ToList();
        return (rows, resp.Seq);
    }

    /// <summary>Creates the server-streaming call object eagerly, in a plain <c>async Task</c>
    /// method, rather than inside a lazy <c>async IAsyncEnumerable</c> iterator -- see
    /// <see cref="ITransport.SubscribeAsync"/>'s doc for why laziness here is a correctness bug,
    /// not just a style nit.
    ///
    /// WHAT THIS DOES AND DOES NOT GUARANTEE, checked against Grpc.Net.Client's actual behavior
    /// (not assumed): creating an <see cref="AsyncServerStreamingCall{TResponse}"/> via the
    /// generated client stub starts the underlying HTTP/2 request immediately -- GrpcCall's
    /// server-streaming path kicks off its send as part of the synchronous prefix of call
    /// creation, it does not wait for the first <c>MoveNextAsync()</c> on the response stream the
    /// way a lazy C# iterator would. So moving call creation out of the iterator (this change)
    /// closes the gap this client itself introduced. What it does NOT give is a hard guarantee
    /// that the SERVER has finished registering the subscription by the time this method returns:
    /// <c>StreamGrpcService.SubscribeTable</c> (Grpc/StreamGrpcService.cs) only calls
    /// <c>stream.SubscribeAsync(...)</c> (the actual Orleans registration) after its handler
    /// starts running, and does not write anything to the response stream -- and ASP.NET Core does
    /// not reliably flush response headers before that either -- until the first delta arrives, so
    /// there is no cheap, hang-free signal to await here that would make this exact. This is the
    /// same best-effort level Python settled for (its grpc.py creates the call object synchronously
    /// and does not wait for any further readiness signal either): the remaining window is a
    /// network round trip plus server-side dispatch, not "however long until this LiveTable's
    /// consumer happens to start reading," which is what the lazy-iterator bug actually cost.</summary>
    public async Task<IAsyncEnumerable<(IReadOnlyList<RowDelta> Deltas, long Seq)>> SubscribeAsync(string table, CancellationToken ct)
    {
        var md = await AuthMetadataAsync(ct).ConfigureAwait(false);
        AsyncServerStreamingCall<TableDeltaBatch> call;
        try
        {
            call = _stream.SubscribeTable(new SubscribeTableRequest { Name = table }, md, cancellationToken: ct);
        }
        catch
        {
            // Nothing to dispose here: SubscribeTable() throwing means no call object was ever
            // created (auth metadata failures surface earlier, above). Kept as a catch/rethrow
            // only to document that this branch was considered, matching SignalRTransport's.
            throw;
        }
        return Consume(call, ct);
    }

    private static async IAsyncEnumerable<(IReadOnlyList<RowDelta> Deltas, long Seq)> Consume(
        AsyncServerStreamingCall<TableDeltaBatch> call, [EnumeratorCancellation] CancellationToken ct)
    {
        try
        {
            await foreach (var batch in call.ResponseStream.ReadAllAsync(ct).ConfigureAwait(false))
            {
                var deltas = batch.Deltas.Select(d => new RowDelta(RowCodec.FromStruct(d.Row), d.Weight)).ToList();
                yield return (deltas, batch.Seq);
            }
        }
        finally
        {
            call.Dispose();
        }
    }

    // ---- ingest ----

    /// <summary>One request, one ack, over a fresh bidi stream -- real HTTP/2 backpressure from
    /// the bidi RPC itself (the server does not ack until the push completes) without holding a
    /// stream open across calls. A long-lived streaming session (sustained backpressure across
    /// many pushes) is future work, not needed for <see cref="StreamsForgeClient.PushAsync"/>.</summary>
    public async Task<IngestAckResult> IngestAsync(
        string source, IReadOnlyList<IReadOnlyDictionary<string, object?>> rows, string? idempotencyKey, bool partial, CancellationToken ct)
    {
        var md = await AuthMetadataAsync(ct).ConfigureAwait(false);
        using var call = _ingest.Ingest(md, cancellationToken: ct);
        var request = new IngestRequest
        {
            SourceName = source,
            Partial = partial,
            IdempotencyKey = idempotencyKey ?? "",
        };
        request.Rows.AddRange(rows.Select(RowCodec.ToStruct));
        await call.RequestStream.WriteAsync(request).ConfigureAwait(false);
        await call.RequestStream.CompleteAsync().ConfigureAwait(false);

        if (!await call.ResponseStream.MoveNext(ct).ConfigureAwait(false))
            throw new StreamsForgeException($"gRPC Ingest('{source}') stream closed with no ack");

        var ack = call.ResponseStream.Current;
        // Thrown HERE rather than left to the caller to notice: the C# codegen's enum member
        // naming (PascalCase, e.g. IngestOutcomeAccepted) is a codegen-convention detail that
        // should not leak into StreamsForgeClient's transport-agnostic error handling -- REST's
        // 202/non-202 split makes the same decision at this same layer, on its own wire's terms.
        if (ack.Outcome != IngestOutcome.Accepted)
        {
            throw new IngestRejectedException(
                string.IsNullOrEmpty(ack.Error) ? $"{source} ingest push rejected: {ack.Outcome}" : ack.Error,
                ack.RowErrors.ToList());
        }
        return new IngestAckResult("INGEST_OUTCOME_ACCEPTED", ack.Accepted, ack.Dropped, ack.Invalid, null, Array.Empty<string>());
    }

    public async ValueTask DisposeAsync()
    {
        await _channel.ShutdownAsync().ConfigureAwait(false);
        _channel.Dispose();
    }
}
